src/constraint_system/scalar_mul.rs

```diff
@@ -0,0 +1,236 @@
+    let g = GENERATOR;
+
+    let g2 = ExtendedPoint::from(GENERATOR) + ExtendedPoint::from(GENERATOR);
+    let g_term = g2 + ExtendedPoint::from(GENERATOR);
+
+    let g3 = AffinePoint::from(g_term);
+
+    ladder[0] = g;
+    ladder[1] = g3;
+    ladder[2] = g.neg();
+    ladder[3] = g3.neg();
+
+    ladder
+}
+
+/// For the purpose of the fixed base add gate, we will be using
+/// value of scalars written in the window NAF form. When a scalar w-NAF 
+/// is evaluated as a w-NAF array, it can only contain values: 1, 3, -1 
+/// or -3.
+/// This calculation for the points within the 'fixed base ladder'
+/// differs depending on the index of the w-NAF that we're evaluating.
+/// The fixed generator point, g, becomes gi, based on the formula:
+/// gi = 4^(n-i)*g. `n` here is the number of elements in the w-NAF
+/// array, where the max is 256. `i` is the round that is currently
+/// being evaluated; where `i` starts at `0` and goes up to n-1.
+pub fn scalar_mul(scalar: Fr, composer: &mut StandardComposer) -> Variable {
+    // Get scalar in correct format
+    // [i8; 256]
+    let mut wnaf_scalar = scalar.compute_windowed_naf(3u8).to_vec();
+    wnaf_scalar.reverse();
+
+    // This is where the accumulator is set.
+    let mut t = Fr::zero();
+    let b = ExtendedPoint::from(GENERATOR);
+    let wnaf_len = wnaf_scalar.len();
+    
+    The initial values
+    // The initial x and y values, aka the offset, differ depending on the 
+    // the conditional sign of the scalar. This is either (4^n + 1) * g, or 
+    // (4^n) * g.
+    if scalar.is_even() {
+        t = Fr::from((4 as usize).pow((wnaf_scalar.len() + 1) as u32) as u64);
+    } else {
+        t = Fr::from((4 as usize).pow(wnaf_scalar.len() as u32) as u64);
+    }
+
+    let mut accums: Vec<Fr> = vec![];
+
+    // The initial accumulator values, aka the offset, differs depending on the 
+    // the conditional sign of the scalar. This is either 1 or (1+1)/4^n. 
+    let mut accum_0 = t;
+    let mut accum_1 = t;
+    accum_0.divn(((4 as usize).pow(wnaf_scalar.len() as u32)) as u32);
+    accum_1.divn(
+        (((4 as usize).pow((wnaf_scalar.len() - 1) as u32))
+            + (wnaf_scalar[wnaf_scalar.len() - 1] as usize)) as u32,
+    );
+
+    accums.push(accum_0);
+    accums.push(accum_1);
+
+    // Using gi from the function description above, we compute 
+    // -gi, 3gi & -3gi. 
+    let mut mul_points: Vec<ExtendedPoint> = vec![];
+    for i in wnaf_scalar.iter().enumerate() {
+        let g_i = ExtendedPoint::from(GENERATOR)
+            * Fr::from((4 as usize).pow((wnaf_scalar.len() - i.0) as u32) as u64);
+        let g_i_neg = g_i.neg();
+        let g_i_3 = g_i * Fr::from(3 as u64);
+        let g_i_neg_3 = g_i_3.neg();
+
+        let g_i_affine = AffinePoint::from(g_i);
+        let g_i_3_affine = AffinePoint::from(g_i_3);
+        let g_i_neg_affine = AffinePoint::from(g_i_neg);
+        let g_i_3_neg_affine = AffinePoint::from(g_i_neg_3);
```

Comment:
Why computing the same points on each iteration instead of just doing it one time and save them?